### ISSUE TITLE
System tray icons disappearing

### ISSUE THREAD

**Issue by andrewdavidwong** [QubesOS/qubes-issues#2242]
#### Qubes OS version (e.g., `R3.1`):

`R3.2-rc2`
#### Problem description:

Multiple users have reported system tray icons "disappearing" while using the i3 window manager:
- https://groups.google.com/d/topic/qubes-users/pRE0yr_EnIQ/discussion
- https://groups.google.com/d/topic/qubes-users/nN7JWaQggZQ/discussion

And under Xfwm4:
- https://github.com/QubesOS/qubes-issues/issues/2242#issuecomment-242303198

And under KDE:
- #2268

---
#### Possibly related issues:
#2216, #2264, #2268

**Comment by cyrinux** [comment]
Hi, I can confirm this too

**Comment by KuroSaru** [comment]
Can confirm the same issue, normally i find it happens mostly if i have to restart i3 window manager.

**Comment by plushambush** [comment]
I can confirm tray icons regularly disappear from notification area when using xfwm4 too.
Skype (mostly), Pidgin, NM, even Qubes manager
3.2rc2

**Comment by andrewdavidwong** [comment]
> I can confirm tray icons regularly disappear from notification area when using xfwm4 too.
> Skype (mostly), Pidgin, NM, even Qubes manager
> 3.2rc2

Thanks; updated. Is that Xfwm4 under Xfce4?

**Comment by adrelanos** [comment]
(debian 8; KDE) Happened to me also during being affected by #2268. During that time the Qubes VM Manger (QVMM) systray was completely gone.

**Comment by plushambush** [comment]
Tthat's right. It is xfwm4 under xfce4

**Comment by tzelch** [comment]
I have noticed a little detail for the problem related to i3.
At least for me only the icons from vms disappear, the Dom0's icon's are always visible, in my case qubes manager and redshift.

**Comment by cyplo** [comment]
On the default Qubes R3.2 installation it happens to me as well, I had the power("battery") icon disappear from tray. I think icons representing panel items stay where they are properly, the problem seems to be happening only to the ones in the small rectangle where Qubes VM Manager, networking and power live by default.

**Comment by andrewdavidwong** [comment]
Related/superset of: #954

**Comment by adrelanos** [comment]
Do the systray icons really disappear? Or are they just transparent? That's easy to confuse, because it's hard to visibly see the transparent systray. Can you check please in the tray area if there is like an invisible item, the one you are missing? If so, the left and right click actions should still work as usual. If so, then this ticket could be a duplicate of #2216.

**Comment by cyrinux** [comment]
For me systray icon become maximized on i3, like a normal App when the
problem appears

Le dim. 12 févr. 2017 17:25, Patrick Schleizer <notifications@github.com> a
écrit :

**Comment by KuroSaru** [comment]
It seams the systray is no longer docked or accessable via the primary
screen. If i restart the qubes agent responsible for it and not the
application the tray icon is fir it will then return. But if i rest i3 or
change any screen settings eg add 2nd screen they disapear again.

**Comment by tzelch** [comment]
Is there any way to support on this issue?
I'm happy to invest some debugging effort if a Dev can tell me where i could start looking.
It really keeps bothering me usability wise.

**Comment by cyrinux** [comment]
Hi guys,
Just for say I can also help to debug this, the problem still here and
really anoying :(
Anybody know if there is same issue with awesome for example?
Regards

**Comment by cyrinux** [comment]
If this can help, I just find a way to fully fix systray and make all app
return to systray.
killall i3bar; shortcut for restart i3; then close all systray window which
are maximized.
I think we can play with a x tool (xprop)? For close automatically all
maximized windows and make a script?

**Comment by resurrexit** [comment]
Like @adrelanos said, nm-applet is just transparent for me. If I hover over the right spot and right/left click, I can use the applet as usual.

**Comment by andrewdavidwong** [comment]
This issue is being closed because:

 - This issue has the "Release 3.2 updates" milestone.
 - [Qubes OS 3.2 recently reached end-of-life (EOL).](https://www.qubes-os.org/news/2019/03/28/qubes-3-2-has-reached-eol/)
 - This issue has not been updated in over a year.

If anyone believes that this issue should be reopened, please let us know in a comment here.

**Comment by adrianx64** [comment]
Hi Andrew,

The issue is not fixed. It is still present on R4.0

**Comment by pkraskov** [comment]
For me this became an issue withing a few days ago after update. (before that I wasn't updated for about 3-4 month). Reproduced constantly with the same behavior  - all qubes icons (except wi-fi) became transparent right after returning from suspended mode.
Test case:
1) reboot OS to enter clear state, see initial set of icons:
![Image](https://i.imgur.com/mZhx91d.png)
2) Suspend OS
3) return from suspend, see whole set of icons for a fraction of a second, then see that all icons except wi-fi gone
![Image](https://i.imgur.com/Fud8PFH.png)
4) hover mouse over empty space and see tooltips, and menu if area clicked which means Icons still there but are became transparent.

Also off topic:  
can somebody tell me why my wi-fi icon looks so strange? it is like it is for about half a year. Are the any easy steps to turn it into more nice look?

**Comment by jevank** [comment]
got this issue after kernel changing 4.14 -> 4.19.43-1. Switching back to 4.14.119-2 version fix the problem.

**Comment by marmarek** [comment]
What GPU do you have @pkraskov @jevank ? Radeon? If that's the case, try 4.19.43-3 (in [testing](https://www.qubes-os.org/doc/testing/#updates) repository).

**Comment by jevank** [comment]
@marmarek  Thinkpad X1 Gen6 with intel GPU

**Comment by jevank** [comment]
In my case icons not disappearing, looks like padding-margin issues. When tuning panel size it comes visible. In addition got black zones on ScreenSaver unlock screen.

**Comment by pkraskov** [comment]
I have same laptop model (Thinkpad x1c6)

**Comment by seberm** [comment]
I have exactly same problem. It started since my last update.  My laptop is Thinkpad T470s.

```
$ lspci | grep VGA
... Intel Corporation HD Graphics 620 (rev 02)
```

```
$ lscpu
... Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz
```

```
$ uname -r
4.19.71-1.pvops.qubes.x86_64
```
Please let me know if you need any more logs.

![qubes-tray](https://user-images.githubusercontent.com/212597/65163033-8d97b980-da3a-11e9-910c-4e1bf6630682.jpeg)

**Comment by marmarek** [comment]
Do you see some issues with dom0 applications? Try dom0 terminal, or settings manager.

**Comment by seberm** [comment]
@marmarek Yes I see, there are sometimes problems with window repainting.

![problem0](https://user-images.githubusercontent.com/212597/65262636-1120db80-db0b-11e9-8925-46326890691f.jpeg)
![problem1](https://user-images.githubusercontent.com/212597/65262637-1120db80-db0b-11e9-8d2b-25962c4cfc48.jpeg)
![problem2](https://user-images.githubusercontent.com/212597/65262638-11b97200-db0b-11e9-94fa-18f60cfd6e57.jpeg)

**Comment by marmarek** [comment]
I can confirm similar issue, but only after switching to text console and back. Just after starting X server (like logout and login back) it works fine, as long as I don't switch to text console.
Does it happen to you @seberm even without switching console?

**Comment by seberm** [comment]
@marmarek unfortunately yes. I am able to reproduce it just with window maximization/minimization.

Strange thing that I can also reproduce it just by selecting some part of my desktop with mouse pointer:

![mouse-problem0](https://user-images.githubusercontent.com/212597/65268433-6d89f800-db17-11e9-9493-2b3917b35fb4.jpeg)

Sometimes I am able to see some crapped content of other windows in these rectangles.

Note that it is **only dom0** where I can see these problems. All the other domains seems to work just fine.

**Comment by jevank** [comment]
> Same behavior after return from suspending. Got this issue after kernel changing 4.14 -> 4.19.43-1. Switching back to 4.14.119-2 version fix the problem.

I can confirm same issues with 4.19.67-1 on thinkpad x1gen6. After switching to fullscreen windows and back. Back to 4.14 again

**Comment by seberm** [comment]
@marmarek 
Correction: It really seems the problem starts after switching to text console and back. After that the problem gets worse and worse over time.

Firstly I thought the problem could be caused by my docking station, but I successfully reproduced it even without the dock.

Now I am running X with this configuration in `/etc/X11/xorg.conf.d/20-intel.conf`:

```
Section "Device"
    Identifier "Intel Graphics"
    Driver "modesetting"
    Option "AccelMethod" "glamor"
    Option "DRI" "3"
EndSection
```

I did some testing and multiple machine reboots and it seems it fixes my problems.

@jevank: Could you please retest with this conf?

**Comment by jevank** [comment]
@seberm this config helps me too, thank you. Will keep working on 4.19 now.

**Comment by oliversturm** [comment]
I have noticed window redrawing issues, but I don't think this has ever happened to dom0 windows. It looks like colorful horizontal lines that cover part of a window, and they go away if I cause the window to redraw by moving it between workspaces or similar. I don't ever switch to a text console during normal work, so that's not related for me. Again, my GPU is a Radeon.

I'm wondering about the thought process behind these questions. I understand that some users have the problem of tray icons not being drawn at all, some others (including myself) see them mangled, but some (again including myself) also see the icons go away entirely - at least that last problem doesn't seem to be connected to GPU trouble as far as I can see. Perhaps there's more than one bug here?

**Comment by anadahz** [comment]
This could be related to the GTK3 bug:
[GtkTrayIcon has no background on X11](https://gitlab.gnome.org/GNOME/gtk/issues/1280)

**Comment by oliversturm** [comment]
Some extra (?) information: I see window redrawing issues in one domain that shows a tray icon which keeps vanishing. This is the sys-net domain that runs nm-applet. I keep a terminal window open for that domain in order to restart nm-applet as needed (to make the tray icon reappear), and this terminal window is frequently covered by white horizontal lines to the point of being unreadable - I can temporarily fix this by shifting the window to a different workspace and back (i3).

On the other hand, a second icon that always vanishes in my setup is that for Skype. The domain Skype runs in never seems to show any drawing issues at all. 

Previous messages in this thread hint at dom0 redrawing issues, and I don't see any of those - the dom0 tray icons always stick around for me as well.

**Comment by sebastianlutter** [comment]
I also have Icons disappearing on i3 wm, the nm-applet and the battery icon. Using qubesOS 4.0.1 on intel i7-8565U with UHD Graphics 620.

They are not invisible, they are not part of the icon bar. The nm-applet is running in sys-net, I have to restart it to add the icon again. I have no idea what process to restart for battery indicator, any hint?

**Comment by weissjeffm** [comment]
I also have disappearing icons in i3, it's not that they're blank, they are just gone and the systray shrinks to fit the remaining icons. The only icons that don't disappear are dom0's. I use external monitors and I think the disappearance may happen when docking in/out (iow running xrandr to turn monitors on an off). I haven't confirmed that yet.

**Comment by SvenSemmler** [comment]
On Mon, Apr 13, 2020 at 08:39:47AM -0700, Jeff Weiss wrote:
> I also have disappearing icons in i3, it's not that they're blank, they are just gone and the systray shrinks to fit the remaining icons. The only icons that don't disappear are dom0's. I use external monitors and I think the disappearance may happen when docking in/out (iow running xrandr to turn monitors on an off). I haven't confirmed that yet.

For me this happens whenever I reload or restart i3.

Mod+Shift+R

/Sven

-- 
 public key: https://www.svensemmler.org/0x8F541FB6.asc
fingerprint: D7CA F2DB 658D 89BC 08D6 A7AA DA6E 167B 8F54 1FB6

**Comment by w1k1n9cc** [comment]
I face the same issue with i3 and R4.0. It also occurs if I change the display settings.

**Comment by jtfidje** [comment]
I too face the same issue if I reload / restart i3, or change display settings (ie, connect a second monitor and use xrandr to extend instead of just mirror the screens).

Using the config given by @seberm did not work for me. Having it in place the tray icons that are missing appears as tiled windows. Closing them makes them appear on the bar, but I cannot "use" them (ie, right click to open context menu)

**Comment by andrewdavidwong** [comment]
Any idea why this might be happening, @marmarta?

**Comment by marmarta** [comment]
@marmarek, look at that - tray icons from VMs disappearing. Are we doing something wrong when monitor count changes?

**Comment by dmoerner** [comment]
I think this bug has recently gotten worse for me on i3. After restarting i3, I am no longer able to even open terminals in vms which had a tray icon killed (sys-net, sys-whonix). The first time I try this, xterm appears for a second and then disappears. After that, I can't get xterm to appear at all. But the vms continue working as far as I can tell.

**Comment by marmarek** [comment]
@dmoerner can you check `/var/log/qubes/guid.sys-net.log` when that happens?

**Comment by dmoerner** [comment]
@marmarek I think I ran into a regression in i3 4.18.2, which I was testing after buliding with qubes-builder. I can't reproduce that extra problem on 4.16-9. With i3 4.16-9, I was able to restart i3 ten times. 7 times, nm-applet appeared as a new window, and when I closed it, it returned to the tray. 3 times, it disappeared, but I could restart it from sys-net. That's the "normal" broken behavior.

With i3 4.18.2, after 3 and then 4 restarts, I was unable to start nm-applet or open anything in sys-net. Here's what was in `/var/log/qubes/guid.sys-net.log`:

```
Icon size: 128x128
ErrorHandler: BadWindow (invalid Window parameter)
                 Major opcode: 10 (X_UnmapWindow)
                 ResourceID:   0x2000121
                 Failed serial number:  3225
                 Current serial number: 3228
  someone else already destroyed this window, ignoring
ErrorHandler: BadWindow (invalid Window parameter)
                 Major opcode: 4 (X_DestroyWindow)
                 ResourceID:   0x2000121
                 Failed serial number:  3226
                 Current serial number: 3228
  someone else already destroyed this window, ignoring
ErrorHandler: BadDrawable (invalid Pixmap or Window parameter)
                 Major opcode: 53 (X_CreatePixmap)
                 ResourceID:   0x200012f
                 Failed serial number:  3357
                 Current serial number: 3359
```
I then found this upstream bug report, https://github.com/i3/i3/issues/4159, and applied that patch. I then got the same problem after 4 restarts, but with a slightly different error message, not sure if this is significant:

```
ErrorHandler: BadWindow (invalid Window parameter)
                 Major opcode: 10 (X_UnmapWindow)
                 ResourceID:   0x20000a4
                 Failed serial number:  1080
                 Current serial number: 1083
  someone else already destroyed this window, ignoring
ErrorHandler: BadWindow (invalid Window parameter)
                 Major opcode: 4 (X_DestroyWindow)
                 ResourceID:   0x20000a4
                 Failed serial number:  1081
                 Current serial number: 1083
  someone else already destroyed this window, ignoring
ErrorHandler: BadDrawable (invalid Pixmap or Window parameter)
                 Major opcode: 72 (X_PutImage)
                 ResourceID:   0x20000b2
                 Failed serial number:  1200
                 Current serial number: 1201
```

**Comment by dmoerner** [comment]
By the way, a temporary way around this problem for people using i3 is to add the following to your i3 config:

```
exec_always --no-startup-id "qvm-run sys-net \"pkill nm-applet; nm-applet \&\""
```
If you have a multihead setup and are facing this problem, you can add a similar qvm-run call to whatever you use to invoke xrandr to switch monitors.

**Comment by qtpies** [comment]
> By the way, a temporary way around this problem for people using i3 is to add the following to your i3 config:
> 
> ```
> exec_always --no-startup-id "qvm-run sys-net \"pkill nm-applet; nm-applet \&\""
> ```
> 
> If you have a multihead setup and are facing this problem, you can add a similar qvm-run call to whatever you use to invoke xrandr to switch monitors.

For me this leads to an icon being shown again, but it isn't clickable

**Comment by ptitdoc** [comment]
I found this reference that was released in 2012 and it seems a good match to the observed issue:
https://i3wm.org/downloads/RELEASE-NOTES-4.5.txt

> i3bar: fake DestroyNotify and send MANAGER ClientMessages to fix tray restarts

Linked issue and commit:
https://github.com/i3/i3/issues/745
https://github.com/i3/i3/commit/e9503a1fa380eec0631696e2cd49eac027f53163

I don't know if it can help troubleshooting the issue.

**Comment by techge** [comment]
Running into this issue as well. As far as I can see, this is fixed in 4.18.3, so the very next release of i3, according to [this issue](https://github.com/i3/i3/issues/4159).

As I am a novice to QubesOS: I guess there is no easy way to test a newer version of i3 as we are highly dependable on fedora-32 repos, is there? I'd gladly give the newer version a shot and test it!

**Comment by anadahz** [comment]
Affects 4.1

**Comment by SaswatPadhi** [comment]
Still affects R4.2. `nm-applet` disappears sometimes, for no apparent reason.

**Comment by UndeadDevel** [comment]
Related issue: #8797

**Comment by ptitdoc** [comment]
> Running into this issue as well. As far as I can see, this is fixed in 4.18.3, so the very next release of i3, according to [this issue](https://github.com/i3/i3/issues/4159).
> 
> As I am a novice to QubesOS: I guess there is no easy way to test a newer version of i3 as we are highly dependable on fedora-32 repos, is there? I'd gladly give the newer version a shot and test it!

As a i3 user, this issue triggers essentially when changing external monitors (usage of xrandr/arandr) as discussed in [this issue](https://github.com/i3/i3/issues/4159).

All systray icons related to my qubes (except dom0 icons) disseapears, but the programs are still alive but both the windows and trayicons disapeared (probably minified but not possible to raise them up anymore because the icon is not there).

Strangely in dom0 i3 is now version 4.22 and this issue is still there.

**Comment by ptitdoc** [comment]
Some experiments...

Listing VMs that are HIDDEN:
```
[dom0 ~]$ for winid in $(xwininfo -tree -root | awk '{print $1}') ; do echo "" ;echo -n "${winid}:" ; xprop -id ${winid} 2>/dev/null|grep _NET_WM
0x36001d8:_NET_WM_STATE(ATOM) = 
0x3a008c1:_NET_WM_STATE(ATOM) = 
0x2a0000c:_NET_WM_STATE(ATOM) = _NET_WM_STATE_FOCUSED
0x3a01441:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3a00c77:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3206820:_NET_WM_STATE(ATOM) = 
0x3206c14:_NET_WM_STATE(ATOM) = 
0x32005f8:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x32001d6:_NET_WM_STATE(ATOM) = 
0x32067c2:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3201044:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x32020ac:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x320097b:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3200298:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x1869b2f:_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE
0x420000c:_NET_WM_STATE(ATOM) = 
0x3c0000c:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3e0000c:_NET_WM_STATE(ATOM) = 
0x3400211:_NET_WM_STATE(ATOM) = 
0x3401bfc:_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
0x3a01e63:_NET_WM_STATE(ATOM) = 
0x2c001d6:_NET_WM_STATE(ATOM) = 
```

Over these VMs, I have different reasons why the VM is hidden. I can check status using `xprop -id VMID`.

Case 1: VM is "normally" hidden because its under another VM in i3wm layout:
```
_NET_WM_USER_TIME(CARDINAL) = 602030681
_NET_WM_DESKTOP(CARDINAL) = 4
_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
WM_STATE(WM_STATE):
		window state: Withdrawn
		icon window: 0x0
_QUBES_VMWINDOWID(WINDOW): window id # 0x2e00007
_QUBES_VMNAME(STRING) = "pentest"
_QUBES_LABEL_COLOR(CARDINAL) = 7688315
_QUBES_LABEL(CARDINAL) = 7
WM_CLASS(STRING) = "pentest:jadx-gui-JadxGUI", "pentest:jadx-gui-JadxGUI"
_NET_WM_ICON(CARDINAL) = 	Icon (128 x 128):
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		user specified location: 0, 0
		program specified location: 0, 0
		program specified minimum size: 200 by 150
WM_COMMAND(STRING) = {  }
WM_ICON_NAME(UTF8_STRING) = "*jadx-gui"
WM_NAME(UTF8_STRING) = "*jadx-gui"
```

Case 2: the application has a tray icon. During sleep mode or monitor change, the application has been minimized but the tray icon disapeared, so I cannot maximize it anymore (or maybe this window is actually the tray icon, I don't know):
```
_NET_WM_USER_TIME(CARDINAL) = 603787153
_NET_WM_DESKTOP(CARDINAL) = 4
_NET_WM_STATE(ATOM) = _NET_WM_STATE_HIDDEN
WM_STATE(WM_STATE):
		window state: Withdrawn
		icon window: 0x0
_QUBES_VMWINDOWID(WINDOW): window id # 0x1200013
_QUBES_VMNAME(STRING) = "pentest"
_QUBES_LABEL_COLOR(CARDINAL) = 7688315
_QUBES_LABEL(CARDINAL) = 7
WM_CLASS(STRING) = "pentest:zim", "pentest:Zim"
_NET_WM_ICON(CARDINAL) = 	Icon (48 x 48):
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified minimum size: 461 by 423
		program specified base size: 0 by 0
WM_COMMAND(STRING) = {  }
WM_ICON_NAME(UTF8_STRING) = "Home:Troubleshoot:Bluetooth - Notebooks"
WM_NAME(UTF8_STRING) = "Home:Troubleshoot:Bluetooth - Notebooks"
```

**Comment by ptitdoc** [comment]
Others experiment: kill i3bar:
`$ killall i3bar`

This will pop all systray icons inclusing hidden one as standard windows.

You can create a now i3bar like this:
`$ i3 --get-socketpath`
`$ i3-msg -t get_bar_config`
`$ i3bar --bar_id mybarid --socket mysocket`

Now how to reattach these trayicons window to the new i3bar ?

**Comment by ptitdoc** [comment]
So normally i3 sends a fake DestroyNotify (https://github.com/i3/i3/blob/b85da284a736edb5c2d629b2880f4fec18bc7575/i3bar/src/xcb.c#L1628). Maybe the message is not propagated to the qube gui daemon ?

**Comment by ptitdoc** [comment]
On the GUI agent side, I see a comment "currently not used":
https://github.com/QubesOS/qubes-gui-agent-linux/blob/02c85aba6f2353fb12dad70b7304ed73807c7267/gui-agent/vmside.c#L2224

Does it means that the event is not propagated properly ? I'm lost in the chain of X11 clients, servers, proxies, agents, daemons.

**Comment by DemiMarie** [comment]
@ptitdoc I think you concluded correctly.

**Comment by FabrizioRomanoGenovese** [comment]
Guys, it's been almost  10 years. Is there any hope to have this fixed in upcoming Qubes 4.3?

**Comment by marmarta** [comment]
> Guys, it's been almost 10 years. Is there any hope to have this fixed in upcoming Qubes 4.3?

i3 is not one of the officially supported window managers, so this has never been very high priority. Hopefully at some point it will work better on wayland. X11 is a bit of a nightmare to debug.

**Comment by andrewdavidwong** [comment]
> i3 is not one of the officially supported window managers, so this has never been very high priority.

According to the issue activity above, it also affects Xfce4, though. (I have no new information on whether this is still the case; just an observation.)

**Comment by FabrizioRomanoGenovese** [comment]
Will wayland be supported on Qubes 4.3?

**Comment by DemiMarie** [comment]
No, R4.4 or later.  I don’t work for ITL anymore, but @valpackett has picked up where my work left off.

**Comment by valpackett** [comment]
..yeah, so that's because R4.3 is already in [release candidate stage](https://www.qubes-os.org/news/2025/09/19/qubes-os-4-3-0-rc2-available-for-testing/) and the Wayland integration work has just started.

Specifically tray icons aren't part of Wayland at all, by the way. For the last decade, the preferred way to provide tray icons —for the DEs that support them at all— has been the [StatusNotifierItem (SNI)](https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/) D-Bus protocol. We might make a bridge for that.

**Comment by DemiMarie** [comment]
Also going to be tricky is tray icon menus.  The `com.canonical.dbusmenu` spec has quite problematic implementations.  COSMIC’s is written in Rust and so hopefully at least somewhat trustworthy, but the rest aren’t.

**Comment by ptitdoc** [comment]
Similar issue in awesome wm when resetting the configuration. But somehow it does not affect when adding a screen like in i3.

In the meantime, I use a script to restart zombie applets explicitly if it helps somebody.

In dom0:
```
dom0:~/.local/share/applications$ cat reset-qubes-applets.desktop 
#!/usr/bin/env xdg-open
[Desktop Entry]
Version=1.0
Type=Application
Terminal=true
Icon=org.xfce.xfdashboard
Name=Reset Qubes Applets
Comment=Reset applets for all running Qubes
Categories=Utility
Exec=/home/user/bin/reset-qubes-applets.sh


dom0:~/.local/share/applications$ cat /home/user/bin/reset-qubes-applets.sh 
#!/bin/bash

for vmname in $(qvm-ls |grep Running|awk '{print $1}' |grep -E -v '(dom0|Name)') ; do
  echo "Checking applets in VM $vmname"
  qvm-run -p $vmname /usr/bin/reset-applets.sh
done
```


In my appvm template (this will require installing xwininfo/xprop commands):
```
user@template:~$ cat /usr/bin/reset-applets.sh 
#!/bin/bash

declare -A RESTART_LIST

# look for all windows with _XEMBED_INFO set
for WIN in $(xwininfo -root -tree|awk '/^\s*0x[0-9a-f]+/ {print $1}'); do
  if xprop -id "$WIN" _XEMBED_INFO 2>/dev/null|grep '=' >/dev/null; then
    # this is an applet, retrieve the pid and the command used to run it
    PID=$(xprop -id "$WIN" _NET_WM_PID|awk '{print $3}')
    COMMAND=$(ps -p "$PID" -o args=)
    echo "Tray/XEmbed window $WIN found with PID $PID ($COMMAND)"
    # add the pid to the list of applet to be restarted
    RESTART_LIST["$PID"]="$COMMAND"
  fi
done

# kill each pid to be restarted and run the applet using the retreived command
for PID in "${!RESTART_LIST[@]}"; do
  COMMAND="${RESTART_LIST[$PID]}"
  echo "Restarting $PID ($COMMAND)"
  kill "$PID"
  nohup $COMMAND&
done
```

This way I just run reset-qubes-applets everytime my screen setup change.

**Comment by ptitdoc** [comment]
(and it also affects Qubes 4.3 by the way)

**Comment by marmarek** [comment]
This is only half-fixed - for GTK apps. Qt apps are still affected. See https://github.com/QubesOS/qubes-gui-daemon/pull/173#issuecomment-3560468259

**Comment by Atrate** [comment]
For me, on `i3` this issue is now completely gone, at least for the apps I use.

**Comment by ben-grande** [comment]
> For me, on `i3` this issue is now completely gone, at least for the apps I use.

I confirm this is fixed on i3. I tried with Qt and Gtk: https://github.com/QubesOS/qubes-gui-daemon/pull/173#issuecomment-3528058583 and they survived some restarts. However, if I do several consecutive i3 restarts, such as 10 restarts in 5 seconds, it breaks. I think this is fine.

>     Qt: SDWDate, KeepassXC
>     GTK: nm-applet, volumeicon

I placed all of these on systray. When the redocking broke, it shows the icon as a window, I closed them and they got back to the tray, the only one that didn't was sdwdate.

---

Can someone test redocking on Xfce or KDE so this issue can be closed?

**Comment by besendorf** [comment]
I never noticed this with KDE since 4.2

**Comment by ben-grande** [comment]
> (and it also affects Qubes 4.3 by the way)

@ptitdoc Can you please try again on Xfce on R4.3?

 @andrewdavidwong If they say it works, this issue can be closed, as it is working on all supported desktop environments, i3, KDE and Xfce.
